### PR TITLE
33Audits [M-01 / M-03]

### PR DESCRIPTION
**Pull request by simplemachine92**
# Description

*What does this PR: do, how, why?*

## Limitations & risks

Addresses M-01, M-03 WIP

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/JBPermissions.sol [hunk 1]
```diff
@@ -55,7 +55,13 @@ contract JBPermissions is IJBPermissions {
     {
         if (permissionId > 255) revert PERMISSION_ID_OUT_OF_BOUNDS();
 
-        return (((permissionsOf[operator][account][projectId] >> permissionId) & 1) == 1);
+        // Always return true if operator has root permission.
+        if (((permissionsOf[operator][account][projectId] >> 1) & 1) == 1) {
```
Comment (mejango) [inline src/JBPermissions.sol:59]: let's use JBPermissionIds.ROOT?

**Comment by mejango** [comment]
@simplemachine92 thoughts on solving the ROOT thing in abstract/JBPermissioned instead of in JBPermissions contract? It's nice to keep the JBPermissions contract returning the literal information on if an operator has a specific permission, regardless if it also has ROOT. 
so something like:

File: src/JBPermissions.sol (hunk 1 above)
Comment (simplemachine92) [reply src/JBPermissions.sol]: yes my fault

**Comment by simplemachine92** [comment]
> @simplemachine92 thoughts on solving the ROOT thing in abstract/JBPermissioned instead of in JBPermissions contract? It's nice to keep the JBPermissions contract returning the literal information on if an operator has a specific permission, regardless if it also has ROOT. so something like: image

I prefer the former solution. ```includeRoot``` is just extra calldata (it's always true in your new code). We could just have a view function that returns ```permissionsOf``` mapping?

**Comment by mejango** [comment]
> > @simplemachine92 thoughts on solving the ROOT thing in abstract/JBPermissioned instead of in JBPermissions contract? It's nice to keep the JBPermissions contract returning the literal information on if an operator has a specific permission, regardless if it also has ROOT. so something like: image
> 
> I prefer the former solution. `includeRoot` is just extra calldata (it's always true in your new code). We could just have a view function that returns `permissionsOf` mapping for the literal info.

true, my thinking was either 1) add extra function that returns the literal info, this is better a boolean flag else the literal info needs to be incoded in a 256 and gives the caller more work to do 2) amend the current function to include flag that determines if the info is literal or ROOT affected, or 3) do not make JBPermissions ROOT aware, instead solve this at the JBPermissioned layer.

**Comment by mejango** [comment]
I took the current solution one step further and added a flag for `includeWildcardProjectId` to remove multiple calls to hasPermissions in cases where we're checking projectId 0. We're now at a case where JBPermissions knows the general ROOT _and_ wildcard projectId schema, thus simplifying life for everyone else. tradeoff is added calldata and logic in JbPermissions.

I'm ok with a different solution and understand where that thought comes from. I think this one here is my preference once tests are passing, it's most gas efficient for our normal cases where both includeRoot and includeWildcardProjectId are true.

**Comment by chinmay-farkya** [comment]
M-01 and M-03 are fixed. But a new bug has been introduced in the hasPermissions() function. In the case that root permission is not given to an operator, the execution will reach the for loop checking all of the specified permissions. It returns false if any of the permissionIds in the list is found to be false. But the problem is it will also return false if all permissions in the list have been granted to the operator. This is because in the happy path, the loop exits without returning anything and the default value of the returned bool type remains false.

**Comment by chinmay-farkya** [comment]
Hmm got it. my mistake idk why but that part of code was not visible to me earlier. Fixed.

**Pull request merged**